acceptance/acceptance_test.go

```diff
@@ -229,1905 +99,1417 @@ func testWithoutSpecificBuilderRequirement(
+						"run_image_mirror":       stackManager.RunImageMirror(),
+						"show_reference":         testLifecycle.ShouldShowReference(),
+						"show_processes":         testLifecycle.ShouldShowProcesses(),
+					},
+				)
+
+				assert.Equal(output, expectedOutput)
+			}
+		})
+
+		it("supports building app from a zip file", func() {
+			stackManager.EnsureDefaultStackCreated()
+			testBuilder := builderManager.EnsureComboBuilderExists()
+			pack.SetDefaultTrustedBuilder(testBuilder.Name())
+
+			appPath := filepath.Join("testdata", "mock_app.zip")
+			output := pack.SuccessfulRunWithOutput("build", appImageName, "-p", appPath)
+
+			assert.NewOutputAssertionManager(output).ReportsSuccessfulImageBuild(appImageName)
+
+			appImageCleanupTask := containerManager.CleanupTaskForImageByName(appImageName)
+			defer appImageCleanupTask()
+		})
+
+		when("builder is untrusted", func() {
+			it("uses the 5 phases", func() {
+				stackManager.EnsureDefaultStackCreated()
+				untrustedBuilder := builderManager.CreateOneOffBuilder()
+				defer containerManager.RemoveImages(untrustedBuilder.Name())
 
-			runImageMirror = value
+				output := pack.SuccessfulRunWithOutput(
+					"build", appImageName,
+					"-p", mockAppPath,
+					"-B", untrustedBuilder.Name(),
+				)
+
+				// TODO: This test could be more explicit by ensuring the correct phase names are included with each message
+				assert.ContainsAll(output, "[detector]", "[analyzer]", "[builder]", "[exporter]")
+				assert.NewOutputAssertionManager(output).ReportsSuccessfulImageBuild(appImageName)
+			})
 		})
 
-		when("creating a windows builder", func() {
+		when("--network", func() {
+			var (
+				tmpDir string
+			)
+
 			it.Before(func() {
-				h.SkipIf(t, dockerHostOS() != "windows", "The current Docker daemon does not support Windows-based containers")
+				h.SkipIf(
+					t,
+					!pack.Supports("build --network"),
+					"--network flag not supported for build",
+				)
+
+				var err error
+				tmpDir, err = ioutil.TempDir("", "archive-buildpacks-")
+				assert.Nil(err)
 			})
 
-			when("experimental is disabled", func() {
-				it("fails", func() {
-					builderName, err := createBuilder(t, runImageMirror, configDir, "", packCreateBuilderPath, lifecyclePath, lifecycleDescriptor)
-					if err != nil {
-						defer h.DockerRmi(dockerCli, builderName)
-						h.AssertError(t, err, "Windows containers support is currently experimental")
-					}
-				})
+			it.After(func() {
+				assert.Nil(os.RemoveAll(tmpDir))
 			})
 
-			when("experimental is enabled", func() {
+			when("the network mode is not provided", func() {
+				it("reports that build and detect are online", func() {
+					stackManager.EnsureDefaultStackCreated()
+					testBuilder := builderManager.EnsureComboBuilderExists()
+					pack.SetDefaultTrustedBuilder(testBuilder.Name())
 
-				var packCreateBuilderHome string
+					buildpackManager.PlaceBuildpacksInDir(tmpDir, components.InternetCapableBuildpack)
```

Comment:
I think you can move these lines to `it.Before`

Reply:
Moreover, the first three lines repeat many times in the tests. Maybe it's worth creating a function for them.

Reply:
This is another of those philosophical decisions. Technically it would be very simple to move these repeated lines into a couple of `Before` functions.

The decision is based on my internalized idea of things I consider **interesting** enough to become part of each test. Each of these things is an important and (potentially) state changing pre-requisite for each of the tests as viewed in isolation.

They're not interesting enough to want to be explicit about how each of them are created, that sort of thing is the **subject** of the test. I do think they're interesting enough that it's worth indicating that these things are happening, and are necessary for the subject to succeed.

Practically this will make it much easier to separate the test suite into multiple files, which is a goal we've discussed previously.

I'd welcome further discussion around this, because it is different to how a lot of tests are written. I'm also thinking about how to externalize my thoughts around the different levels of interesting that exist in a test suite, and how we can organize them most effectively.

Reply:
Thank you for your detailed answer!
From my point of view, although these lines are important and interesting, I think that duplicate code is really bad.
If you want to change the behavior or you find a bug, you need to change it in multiple places. Moreover, contributors that add new tests might forget to add these lines.
Anywary, I totally agree that think it's worth having further discussion around this.